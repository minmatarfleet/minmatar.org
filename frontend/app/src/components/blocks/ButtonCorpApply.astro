---
import { i18n } from '@helpers/i18n'
const { t, translatePath } = i18n(Astro.url)

import type { CorporationObject } from '@dtypes/layout_components'

interface Props {
    corporation:            CorporationObject;
}

const {
    corporation,
} = Astro.props

const CORPORATION_REQUEST_STATUS_PARTIAL_URL = `${translatePath('/partials/corporation_slide_request_status_item/')}?corporation_id=${corporation.corporation_id}`

const REAPPLY_PERIOD_HOURS = 1
let reapplication_date = new Date(corporation?.application_updated as Date)
reapplication_date.setHours(reapplication_date.getHours() + REAPPLY_PERIOD_HOURS);

const MAX_APPLICATION_TEXT_LENGTH = 2000

import Flexblock from '@components/compositions/Flexblock.astro';
import FlexInline from '@components/compositions/FlexInline.astro';

import DialogButton from "@components/blocks/DialogButton.astro";
import Button from '@components/blocks/Button.astro';
import CorporationItem from '@components/blocks/CorporationItem.astro';
import Textarea from '@components/blocks/Textarea.astro';

import Markdown from '@components/icons/buttons/Markdown.astro';
---

<DialogButton
    size='sm'
    color='green'
>
    {t('enlist_today')}

    <form
        slot="dialog" 
        hx-post={CORPORATION_REQUEST_STATUS_PARTIAL_URL}
        hx-trigger="submit"
        hx-target={`#corporation-status-${corporation.corporation_id}`}
        hx-indicator=".loader"
        hx-swap="outerHTML transition:true"
        x-data={`{
            application_text: '',
            error: false,
            clear_form() {
                $nextTick(() => {
                    this.application_text = ''
                })
            }
        }`}
        x-on:submit="close_dialog(); clear_form()"
    >
        <Flexblock gap="var(--space-l)">
            <Flexblock gap="var(--space-s)">
                <h3>{t('corporation_join_request_dialog_title')}</h3>
                
                <p>{t('corporation_join_request_dialog_text')}</p>
                <CorporationItem corporation={corporation} disable_applications={true} />
                <Flexblock class="[ w-full ]" gap="var(--space-3xs)">
                    <label for="description">{t('application_text')}</label>
                    <div class="[ remain-wrapper ]" x-data={`{ limit: $el.dataset.limit }`} data-limit={MAX_APPLICATION_TEXT_LENGTH}>
                        <Textarea
                             x-ref="content"
                             id="description"
                             name="application_text"
                             x-model="application_text"
                             rows="5"
                             x-on:input={`error = !application_text || application_text.length > ${MAX_APPLICATION_TEXT_LENGTH}`}
                             required
                        />
                        <small
                            x-show={`application_text.length > ${MAX_APPLICATION_TEXT_LENGTH - 200}`}
                            x-ref="remaining"
                            x-text="limit - application_text.length"
                            x-bind:class="(error ? 'text-red animate-fade-in-up' : '')"
                        >
                            {MAX_APPLICATION_TEXT_LENGTH}
                        </small>
                    </div>
                    <a href="https://www.markdownguide.org/" class="[ w-fit ]" target="_blank" rel="noreferrer noopener nofollow">
                        <FlexInline gap='var(--space-3xs)'>
                            <Markdown />
                            <small>{t('markdown_support')}</small>
                        </FlexInline>
                    </a>
                </Flexblock>
            </Flexblock>

            <FlexInline justification='flex-end'>
                <Button
                    size='sm'
                    type="submit"
                    color='green'
                >
                    {t('apply')}
                </Button>
                <Button
                    type="button"
                    size='sm'
                    x-on:click="close_dialog()"
                >
                    {t('close')}
                </Button>
            </FlexInline>
        </Flexblock>
    </form>
</DialogButton>

<style lang="scss">
    .remain-wrapper {
        position: relative;

        small {
            position: absolute;
            right: var(--space-2xs);
            bottom: var(--space-3xs);

            &.text-red {
                font-weight: 600;
            }
        }
    }
</style>