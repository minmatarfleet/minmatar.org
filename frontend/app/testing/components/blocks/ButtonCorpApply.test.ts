import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test } from "vitest";
import ButtonCorpApply from "@components/blocks/ButtonCorpApply.astro";

test("ButtonCorpApply defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(ButtonCorpApply, {
    props: {
      corporation: {
        application_updated: new Date(),
        corporation_id: 8324,
      },
    },
  });

  expect(result).toMatchSnapshot();
});
