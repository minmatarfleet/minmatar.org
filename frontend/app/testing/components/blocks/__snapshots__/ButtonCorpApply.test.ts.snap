// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ButtonCorpApply defaults 1`] = `
"<div x-data="{
        button_dialog_open: false,
        open_dialog() {
            this.button_dialog_open = true
        },
        close_dialog() {
            this.button_dialog_open = false
        }
    }" data-astro-cid-xwghkimo> <button type="button" x-on:click.stop.prevent="open_dialog()" data-astro-cid-xwghkimo="true" data-size="sm" style="--button_background_color: var(--green); --button_color: var(--fleet-yellow)" disabled x-init="$el.removeAttribute(&#34;disabled&#34;)" data-astro-cid-r5e2wo56="true" class="[ button ]"> <div class="flex-inline" data-astro-cid-r5e2wo56="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-xs);--flex_inline_justification: center;">    Submit Application    </div>   </button>  <div role="dialog" class="dialog solid-transparency breakout non-centered fixed button-dialog w-full component overflow-y-auto max-w-[700px]" x-bind:class="(button_dialog_open ? 'open' : '')" x-bind:aria-hidden="button_dialog_open == false" x-trap="button_dialog_open" x-bind:inert="!button_dialog_open" @keyup.escape.window="close_dialog()" @mousedown.outside="open &#38;&#38; close_dialog()" data-astro-cid-xwghkimo="true"> <div class="wrapper" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-m);--wrapper_padding_inline_end: var(--space-m);--wrapper_padding_block: var(--space-m);--wrapper_max_width: 80rem;">  <div class="flex-block gap---space-m" data-mod style="--flex_block_gap: var(--space-m);">   <div class="[ cover-body ]">  <form hx-post="/partials/corporation_slide_request_status_item/?corporation_id=8324" hx-trigger="submit" hx-target="#corporation-status-8324" hx-indicator=".loader" hx-swap="outerHTML transition:true" x-data="{
            application_text: '',
            error: false,
            clear_form() {
                $nextTick(() => {
                    this.application_text = ''
                })
            }
        }" x-on:submit="close_dialog(); clear_form()" data-astro-cid-gnhukzia> <div class="flex-block gap---space-l" data-astro-cid-gnhukzia="true" data-mod style="--flex_block_gap: var(--space-l);">  <div class="flex-block gap---space-s" data-astro-cid-gnhukzia="true" data-mod style="--flex_block_gap: var(--space-s);">  <h3 data-astro-cid-gnhukzia>Corporation join request</h3> <p data-astro-cid-gnhukzia>Welcome to the first step of your glorious descent into organized chaos. Fill out your application text to submit your application to our corporation?</p> <div class="fixed-fluid [ corporation-item items-center ]" id="corporation-item-8324" x-data="{
        corporation_item_init() {
            
        }
    }" x-init="corporation_item_init()" data-astro-cid-cn6tudgz="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 64px;--fixed_fluid_breakpoint: 50%;">  <picture class="[ border ]" data-astro-cid-cn6tudgz> <img loading="lazy" src="https://images.evetech.net/corporations/8324/logo?size=64" width="64" height="64" alt="undefined corporation logo" data-astro-cid-cn6tudgz> </picture> <div class="flex-inline [ items-center ]" data-astro-cid-cn6tudgz="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: space-between;">  <div class="flex-block [ grow basis-[220px] ] gap-0" data-astro-cid-cn6tudgz="true" data-mod style="--flex_block_gap: 0;">  <span class="[ text-[var(--highlight)] ]" data-astro-cid-cn6tudgz>  </span>  </div>  <div id="corporation-status-8324" x-data="{
        init() {
            
        }
    }">  </div>  </div>    </div>    <div class="flex-block [ w-full ] gap---space-3xs" data-astro-cid-gnhukzia="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <label for="description" data-astro-cid-gnhukzia>Application text</label> <div class="[ remain-wrapper ]" x-data="{ limit: $el.dataset.limit }" data-limit="2000" data-astro-cid-gnhukzia> <textarea x-ref="content" id="description" name="application_text" x-model="application_text" rows="5" x-on:input="error = !application_text || application_text.length > 2000" required data-astro-cid-gnhukzia="true" data-astro-cid-2vffnk23="true"></textarea>  <small x-show="application_text.length > 1800" x-ref="remaining" x-text="limit - application_text.length" x-bind:class="(error ? 'text-red animate-fade-in-up' : '')" data-astro-cid-gnhukzia> 2000 </small> </div> <a href="https://www.markdownguide.org/" class="[ w-fit ]" target="_blank" rel="noreferrer noopener nofollow" data-astro-cid-gnhukzia> <div class="flex-inline" data-astro-cid-gnhukzia="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-3xs);--flex_inline_justification: flex-start;">  <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"> <path d="M16 15L19 12L17.95 10.925L16.75 12.125V9H15.25V12.125L14.05 10.925L13 12L16 15ZM2 20V4H22V20H2ZM5.5 15H7V10.5H8V13.5H9.5V10.5H10.5V15H12V9H5.5V15Z" fill="currentColor"></path> </svg> <small data-astro-cid-gnhukzia>This field supports Markdown</small>  </div>   </a>  </div>   </div>  <div class="flex-inline" data-astro-cid-gnhukzia="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: flex-end;">  <button type="submit" data-astro-cid-gnhukzia="true" data-size="sm" style="--button_background_color: var(--green); --button_color: var(--fleet-yellow)" disabled x-init="$el.removeAttribute(&#34;disabled&#34;)" data-astro-cid-r5e2wo56="true" class="[ button ]"> <div class="flex-inline" data-astro-cid-r5e2wo56="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-xs);--flex_inline_justification: center;">   Apply  </div>   </button>  <button type="button" x-on:click="close_dialog()" data-astro-cid-gnhukzia="true" data-size="sm" style="--button_background_color: var(--fleet-red); --button_color: var(--fleet-yellow)" disabled x-init="$el.removeAttribute(&#34;disabled&#34;)" data-astro-cid-r5e2wo56="true" class="[ button ]"> <div class="flex-inline" data-astro-cid-r5e2wo56="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-xs);--flex_inline_justification: center;">   Close  </div>   </button>   </div>    </div>  </form>  </div>   </div>   </div>   </div>  </div>  "
`;
